### DISCUSSION TITLE
A perspective from a multimedia company

### DISCUSSION THREAD

**Discussion by brainw0rm** [cables-gl/cables_docs#329] [Feature Requests]
Hello everyone!

First of all I want to say thank you to the amazing people who are developing and accompanying this amazing journey that is cables.
I personally fell in love with it a few months ago and am trying to implement it more and more into my professional workflow.

To give some background: Our company is called Responsive Spaces and we are a B2B marketing agency based in Austria which focuses mostly on interactive multimedia installations in showrooms, trade fairs, exhibitions and so on. We have been mostly using vvvv for our project setups as it provides a lot of flexibility when it comes to the field we are operating in.

Now I am at a point where I start to think in cables rather than other applications when it comes to designing new concepts and content for new projects. This led me and my colleagues to identify some major drawbacks (related to the nature of our projects rather than cables of course) that most of the time hinder us in using cables effectively as one of our preferred tools for realizing ideas.

I will try to shortly summarize them in the following and want to maybe hear other people's thoughts about it and get a bit of reflection on whether these topics are also relevant in a broader scope or if we are just hitting a niche here which has absolutely no priority in the general cables environment. I apologize in advance if some of these things are considered totally absurd inside the cables framework as I am not fully aware its potential capabilities.

Topics would be (in descending priority):

1) It is often the case that we need to use/split our content to multiple screens/beamers. We basically need to be able to have multiple renderers/viewports in one single project where we can choose which textures are viewed on which renderer/viewport.

2) As we often work with high resolution texture/video-texture content which is stored and exchanged locally on our servers/pcs, having no way of accessing data locally on the same pc in realtime (i.e. reading and playing videos from a local folder) makes things very complicated to set up for us.

3) It is often necessary for us to implement some kind of sensor data into our patches which often come via serial or udp packets. Having some kind of native nodes for serial communication or receiving udp would vastly improve our flexibility in interactive use cases.

Again, the purpose of this thread is only to maybe get some thoughts and reflections on these topics and are not meant in any way as criticism nor should it be considered as a classical "feature request" thread. It is just what has been on our minds over the last few weeks/month when thinking about one of our new favorite tools :)

Thank you for reading and have a very nice day everyone!

**Comment by steam0r** [comment]
hey,

thanks for your kind words and starting this talk. we would also be interested in how people approach these situations, as they are mostly "problems" that come from the fact that cables is running in a browser and hence has to adhere to all the sandboxing and security measures that come with that environment.

this goes for multiple renderers (would need some protocol between browserwindows), storing files locally (which can be workarounded with using virtual webcams, starting browser without security features (which became even harder to do lately) or running some local proxy (which is a beast on it's own) and also providing sensor data to cables via serial and udp (which can also be done with external tools, additional proxies, ...).

so maybe we use this tread to collect that people have found or done to make cables more suited for "installation environments". 

looking forward to it!

**Reply by steam0r** [reply]
some hints, all of this is to be taken with a grain of salt and "as is":

allow insecure websocket requests in browsers: 
https://www.damirscorner.com/blog/posts/20210528-AllowingInsecureWebsocketConnections.html

osc to websocket: 
https://github.com/pandrr/osc2ws

cables daplaya: 
https://github.com/cables-gl/cables-daplaya

cables electron export:
https://cables.gl/docs/4_export_embed/dev_embed/export_exe/export_exe

insecure browser: 
https://github.com/pandrr/insecurebrowser

**Reply by Rufussed** [reply]
Steamor I tried your Cables-Daplaya and it works great for making a standalone offline patch player.  Thanks for your work on that - amazing.  Bionic Flamingo it also provides a solution for using local media files.  I included a image asset in my patch.  It was downloaded and stored in an Assets folder when Daplaya downloaded the patch for offline use.  I was then able to simply rename another image file with the same name and replace it in the assets folder, and it ran with the replaced asset without issue.

**Comment by pandrr** [comment]
thanks for your thoughts!
undev can also always be booked to patch, help and consult ... we can hop on a call if interested!
working in the browser environment has its pro and cons vs native software

**Comment by brainw0rm** [comment]
Hi guys!
Thx for the comments and the very interesting links! Yeah I totally understand that many of the limitations come from the very nature of operating in a browser environment. Thats also what makes it very appealing on the other hand.

I had a few thoughts on these topics:

Wouldnt it be possible to maybe do something like multiple remote viewing tabs like the one which is already available and have like an op that takes a texture as an input and can route multiple of them to different tabs? That way it could be possible to just put a different tab on each different display?

As for not having access to local files on the pc: I think it would be probably a very good option to offer something like a pro subscription which lifts the limitation of the webserver and lets you upload bigger files than 100mb and give some kind of predefined amount of space you can fill up with your project files. This could also be on a per project quota. Like paying a fee for a single patch which then disables the upload limitations.

We will definitely reach out to undev if we decide to realize a bigger project in cables and need some guidance, thx very much for the offer!

I wish you all the best and have a very nice day!

**Reply by jimimased** [reply]
"a pro subscription which lifts the limitation of the webserver and lets you upload bigger files than 100mb" is a very good suggestion & much needed in my opinion.
